### ISSUE TITLE
Cannot get certificate

### ISSUE THREAD

**Issue by zoc** [freswa/dovecot-xaps-daemon#4]
Hi, I've been trying to update from the original xapsd implementation from st3fan today to yours (because it seems you're still maintaining it and it would avoid me from keeping a very old OSX VM image just to renew the certs...).

Unfortunately it seems xapsd is not able to get a certificate, at least for now. I followed the guidelines for the password (the account has double auth enabled, could it be an issue ?), and ended up adding some debug logs into your Go source code (I'm actually not very familiar with Go, but as a day to day .NET C# and F# senior dev I think I can deal with it :) ).

First time (**only**) I launched xapsd I got this error

> Dec 19 10:53:54 mail.zoc.me xapsd[23671]: 2020/12/19 10:53:54 Error -20358 while retrieving certificates:
> Dec 19 10:53:54 mail.zoc.me xapsd[23671]: {Response:{Status:{ErrorDescription:Exception from DS ErrorMessage:Invalid Session ErrorCode:-20358} Certificates:[]} Header:{ClientIPAddress:1 LanguagePreference:1 TransactionId:1 ClientOSVersion:2.1 ClientOSName:MAC OSX ClientApplicationName:XServer ClientApplicationCredential:1}}

But since then no clear error messages (even after deleting the xapsd db to restart from a clean state).

So, basically I decided to dump the response from Apple when submitting certs request and this is what I get:
> 2020/12/19 12:11:20 \<html>
> \<head><title>502 Bad Gateway\</title>\</head>
> \<body>
> \<center>\<h1>502 Bad Gateway\</h1>\</center>
> \<hr>\<center>Apple\</center>
> \</body>
> \</html> 

Then the process fails with this:
> 2020/12/19 12:11:20 plist: error parsing text property list: unexpected hex digit `h' at line 0 character 2`

which I guess is "expected" given what the parser gets as input....

Have you ever seen this from Apple ? 
Could I've been blacklisted from my first attempt ?
Did you succeed to get certificates using xapsd recently ?

Thanks for your work and time.

**Comment by freswa** [comment]
> Have you ever seen this from Apple?
Yes, they sometimes do maintenance for days. Since the domain is still available but their upstream is offline this can be the case again.

> Could I've been blacklisted from my first attempt?
Could be, but I haven't seen this before.

> Did you succeed to get certificates using xapsd recently?
I did.

If this will start to work again, I'll add some checks for the status code and appropriate error messages.

**Comment by zoc** [comment]
Thanks for the feedback. I'll give it a try in a few days to see if Apple servers are back online.

**Comment by freswa** [comment]
Seems it's up again.

**Comment by zoc** [comment]
Correct, it's back online, but still not available to get a certificate:

> Dec 22 12:35:35 mail.zoc.me xapsd[17623]: {Response:{Status:{ErrorDescription:Exception from DS ErrorMessage:Invalid Session ErrorCode:-20358} Certificates:[]} Header:{ClientIPAddress:1 LanguagePreference:1 TransactionId:1 ClientOSVersion:2.1 ClientOSName:MAC OSX ClientApplicationName:XServer ClientApplicationCredential:1}}

Wrong password maybe ?

I used the following when creating the password hash, because for some weird reasons I had to press Ctrl+D three times instead of two and I wanted to avoid potential issues:

`echo -n 'MyAppleIDPasswordHere' | openssl dgst -sha256 -binary | xxd -p -c 32`

To me the result should be the same. 

Do you have 2FA enabled on the account you're using to retrieve your certificates ?

Don't really know what to do next. Any advice ?

**Comment by freswa** [comment]
This is quite tricky and I don't like the current scheme. The output the command echoes, overwrites the output line and thus that very line includes not only the hash but also the password. So maybe you not only copied the hash but also the password?

2FA is not an issue.

**Comment by zoc** [comment]
I noticed the password in clear text at the beginning of the result, and I made sure to copy only the hash.

**Comment by freswa** [comment]
I'm getting 502 again :/

**Comment by zoc** [comment]
Still working for me, but same result...

I actually also tried with https://github.com/scintill/macos-server-apns-certs : Same result, does not work, same error returned.

**Comment by freswa** [comment]
Looks like they are doing some round robin stuff and one of their upstreams is offline.
I added some code to catch HTTP errors.

I have no idea what this error is about. Maybe @scintill has a clue?

Just for the reference:
```
2020/12/22 18:13:37 Error -20358 while retrieving certificates:
{Response:{Status:{ErrorDescription:Exception from DS ErrorMessage:Invalid Session ErrorCode:-20358} Certificates:[]} Header:{ClientIPAddress:1 LanguagePreference:1 TransactionId:1 ClientOSVersion:2.1 ClientOSName:MAC OSX ClientApplicationName:XServer ClientApplicationCredential:1}}
```

**Comment by zoc** [comment]
Hi,

More on this topic: I'm not even able to get a certificate when using my macOS Sierra VM. Same error code : -20358

I don't know if there's something wrong with my apple id. I'll try to create a new one and see what happens.

**Comment by bconfais** [comment]
Also, some existing certificates do not work anymore: push.apple.com replies that the issuer is incorrect.

**Comment by freswa** [comment]
I can't make it work on any macOS system with multiple different Apple IDs either.

**Comment by scintill** [comment]
I was able to get new certificates 2.5 hours ago (using https://github.com/scintill/macos-server-apns-certs/) Maybe the Apple service was just broken for awhile... can someone else try again?

The push notifications seem to still be working with my current cert too (granted, I only have one client since it's my personal server.)

**Comment by zoc** [comment]
Same for me actually just now using xapsd. That's maybe Apple's way to celebrate new year ;)

Now it's my turn: Happy new year and all the best to all of you.

**Comment by scintill** [comment]
Good to hear @zoc! Happy new year to you all. :)

If others still have problems, maybe it is a round robin like @freswa mentioned, so you may have to keep trying until you get a server that works.

**Comment by bconfais** [comment]
I am able to create the certificate but when the push notification is sent to apple server, I got the following error:

`Jan 01 06:45:54 srv xapsd[30775]: time="2021-01-01T06:45:54Z" level=fatal msg="Error:Post \"https://api.push.apple.com/3/device/5A0FF26D0A43A53133B3F3257FA42F6432BC945D908E5FFEA3BBA2D229CFBF13\": x509: certificate signed by unknown authority"`
